File: PracticaExamen/AccesoDatos/DAOFactura.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Data.SqlClient;
using System.Data;

using ObjetosTransferencia;

namespace AccesoDatos
{
    public class DAOFactura
    {
        private ConnectionManager connection;
        private SqlDataAdapter adaptador;

        public DAOFactura() {
            connection = new ConnectionManager();
            adaptador = new SqlDataAdapter();

            adaptador.InsertCommand = crearInsert();
        }

        public TOFactura guardar(TOFactura factura) {

            String query = "Insert into Factura Values(@numero, GETDATE(), @cedula, @total);";
            SqlCommand c = new SqlCommand(query, connection.connection);

            int cod = (this.obtenerCodigo() + 1);
            factura.Detalle.numeroFactura = cod;

            c.Parameters.AddWithValue("@numero", cod );
            c.Parameters.AddWithValue("@cedula", factura.Cliente.Cedula);
            c.Parameters.AddWithValue("@total", factura.Total);
            
            try
            {
                connection.abrir();
                c.ExecuteNonQuery();
                connection.cerrar();

                guardarProductos(factura.Detalle.lista);
            }
            catch (SqlException ex)
            {

                throw new Exception(ex.Message);
            }
            finally {
                connection.cerrar();
            }

            factura.ID_Factura = cod;
            return factura;
        }

        private void guardarProductos(List<TOProducto> list) {
            try
            {
                DataTable d = crearDataTable(list);
                adaptador.Update(d);
            }
            catch (SqlException ex)
            {

                throw new Exception(ex.Message);
            }
        }

        private DataTable crearDataTable(List<TOProducto> list) {

            DataTable dataTable = new DataTable();
            dataTable.Columns.Add("numero");
            dataTable.Columns.Add("cod");
            dataTable.Columns.Add("des");
            dataTable.Columns.Add("precio");

            int cod = this.obtenerCodigo();

            foreach (TOProducto item in list)
            {
                dataTable.Rows.Add(cod, item.Codigo, item.Descripcion, item.Precio);
            }

            return dataTable;
        }

        private SqlCommand crearInsert() {

            String query = "INSERT INTO DetalleFactura VALUES(@num, @cod, @des, @precio);";

            SqlCommand c = new SqlCommand(query, connection.connection);
            SqlParameter num = new SqlParameter("@num", SqlDbType.NChar, 10, "numero");
            SqlParameter cod = new SqlParameter("@cod", SqlDbType.NVarChar, 50, "cod");
            SqlParameter des = new SqlParameter("@des", SqlDbType.NVarChar, 200, "des");
            SqlParameter precio = new SqlParameter("@precio", SqlDbType.Money, 10, "precio");

            c.Parameters.Add(num);
            c.Parameters.Add(cod);
            c.Parameters.Add(des);
            c.Parameters.Add(precio);

            return c;
        }

        private SqlCommand obtenerUltimoCodigo() {
            String query = "SELECT Count(*) From Factura;";
            SqlCommand c = new SqlCommand(query, connection.connection);

            return c;
        }

        private int obtenerCodigo() {
            SqlCommand c = this.obtenerUltimoCodigo();
            int count = 0;
            try
            {
                connection.abrir();
                count = Int32.Parse(c.ExecuteScalar().ToString());
                return count;
            }
            catch (Exception ex)
            {

                throw new Exception(ex.Message);
            }
            finally {
                connection.cerrar();
            }
        }
    }
}

File: Practicav2/AccesoDatos/DAOFactura.vb
Imports ObjetosTransferencia
Imports System.Data
Imports System.Data.SqlClient

Public Class DAOFactura
    Dim conex As ConnectionManager
    Dim comando As SqlCommand

    Sub New()
        conex = New ConnectionManager
    End Sub

    Function guardar(f As TOFactura) As TOFactura

        Dim daoD As DAODetalleFactura = New DAODetalleFactura
        Dim num As Integer = getNum() + 1
        f.Numero = num
        f.detalle.NumeroFactura = num

        conex.abrir()
        comando = Me.insertCommand(f)

        comando.ExecuteNonQuery()

        conex.cerrar()

        daoD.guardarDetalle(f.detalle)

        Return f
    End Function

    Private Function insertCommand(f As TOFactura) As SqlCommand

        Dim query As String = "Insert into Factura Values(@num, GETDATE(), @ced, @total)"

        Dim c As SqlCommand = New SqlCommand(query, conex.connection)
        c.Parameters.AddWithValue("num", f.Numero)
        c.Parameters.AddWithValue("ced", f.ClienteF.Cedula)
        c.Parameters.AddWithValue("total", f.Total)

        Return c
    End Function

    Private Function getNum() As Integer
        conex.abrir()
        comando = Me.countCommand
        Dim num As Integer = comando.ExecuteScalar()
        conex.cerrar()
        Return num
    End Function

    Private Function countCommand() As SqlCommand
        Dim query As String = "Select Count(*) From Factura"
        Dim c As SqlCommand = New SqlCommand(query, conex.connection)

        Return c
    End Function

End Class
